libs/frontend/feature-build/src/lib/prosemirror-highlighthjs-plugin.ts

```diff
@@ -0,0 +1,274 @@
+): Plugin<HighlightPluginState> {
+  const extractor =
+    languageExtractor ||
+    function (node: ProseMirrorNode) {
+      const detectedLanguage = node.attrs.detectedHighlightLanguage as string
+      const params = node.attrs.params as string
+      return detectedLanguage || params?.split(' ')[0] || ''
+    }
+
+  // const setter =
+  //   languageSetter ||
+  //   function (tr, node, pos, language) {
+  //     const attrs = node.attrs || {}
+
+  //     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  //     // @ts-expect-error
+  //     attrs['detectedHighlightLanguage'] = language
+
+  //     // set the params attribute of the node to the detected language
+  //     return tr.setNodeMarkup(pos, undefined, attrs)
+  //   }
+
+  const getDecos = (doc: ProseMirrorNode, cache: DecorationCache) => {
+    const autodetectedLanguages: {
+      node: ProseMirrorNode
+      pos: number
+      language: string | undefined
+    }[] = []
+
+    const content = getHighlightDecorations(doc, hljs, nodeTypes, extractor, {
+      preRenderer: (_, pos) => cache.get(pos)?.decorations,
+      postRenderer: (b, pos, decorations) => {
+        cache.set(pos, b, decorations)
+      },
+      autohighlightCallback: (node, pos, language) => {
+        autodetectedLanguages.push({
+          node,
+          pos,
+          language,
+        })
+      },
+    })
+
+    return { content, autodetectedLanguages }
+  }
+
+  // key the plugin so we can easily find it in the state later
+  const key = new PluginKey<HighlightPluginState>()
+
+  return new Plugin<HighlightPluginState>({
+    key,
+    state: {
+      init(_, instance) {
+        const cache = new DecorationCache({})
+        const result = getDecos(instance.doc, cache)
+        return {
+          cache: cache,
+          decorations: DecorationSet.create(instance.doc, result.content),
+          autodetectedLanguages: result.autodetectedLanguages,
+        }
+      },
+      apply(tr, data) {
+        const updatedCache = data.cache.invalidate(tr)
+        if (!tr.docChanged) {
+          return {
+            cache: updatedCache,
+            decorations: data.decorations.map(tr.mapping, tr.doc),
+            autodetectedLanguages: [],
+          }
+        }
+
+        const result = getDecos(tr.doc, updatedCache)
+
+        return {
+          cache: updatedCache,
+          decorations: DecorationSet.create(tr.doc, result.content),
+          autodetectedLanguages: result.autodetectedLanguages,
+        }
+      },
+    },
+    props: {
+      decorations(this: Plugin<HighlightPluginState>, state) {
+        return this.getState(state)?.decorations
+      },
+    },
+    // `prosemirror-highlightjs` use this function to update node's language in a editor,
+    // which leads to a "Maximum call stack size exceeded" error. Since we only need to
+    // format JS code at the moment, we can just skip the following part for now.
+
+    // view(initialView: EditorView) {
```

Comment:
This `view` spec is what causes the "Maximum call stack size exceeded" error. From the code it seems that it's used to update the language of the code for highlightjs. But since we only care about JavaScript at the moment, I just commnted out the whole section to make it work.

Reply:
I guess it probably would have been sufficient to pass in a blank/empty `languageSetter` function when initializing the plugin, but disabling the code is clear.